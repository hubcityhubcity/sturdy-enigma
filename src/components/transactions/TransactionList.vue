<template>
  <div>
    <template v-for="group in groupedTransactions">
      <div :key="group[0].title">
        <h3>{{ group[0].title }}</h3>
        <TransactionItem
          v-for="tx in group"
          :key="tx.hash"
          :transaction="tx.tx"
          :validators="validators"
          :address="address"
        />
      </div>
    </template>
  </div>
</template>

<script>
import TransactionItem from "./TransactionItem"
import groupBy from "lodash.groupby"
import orderBy from "lodash.orderby"
import moment from "moment"

const categories = [
  {
    title: "Today",
    matcher: tx => {
      return moment(tx.timestamp).isSame(moment(), "day")
    }
  },
  {
    title: "Yesterday",
    matcher: tx => {
      return moment(tx.timestamp).isSame(moment().subtract(1, "days"), "day")
    }
  }
]

export default {
  name: `transaction-list`,
  components: {
    TransactionItem
  },
  props: {
    transactions: {
      type: Array,
      required: true
    },
    address: {
      type: String,
      default: null
    },
    validators: {
      type: Object,
      required: true
    }
  },
  computed: {
    groupedTransactions() {
      return orderBy(
        groupBy(this.categorizedTransactions, "title"),
        group => group[0].tx.timestamp,
        "desc"
      )
    },
    categorizedTransactions() {
      return this.transactions.map(tx => {
        // check if the tx is in Today, Yesterday or Last Week
        const dateString = ` (` + moment(tx.timestamp).format("MMMM Do") + `)`
        const category = categories.find(({ matcher }) => matcher(tx))
        if (category) {
          return {
            title: category.title + dateString,
            tx
          }
        }

        // check if tx is in a month this year
        const date = moment(tx.timestamp)
        const today = moment()
        if (date.year() === today.year()) {
          return {
            title: date.format("MMMM Do"),
            tx
          }
        }

        // tx is in a month another year
        return {
          title: date.format("MMMM Do, YYYY"),
          tx
        }
      })
    }
  }
}
</script>
<style scoped>
h3 {
  margin: 2rem 0 0.5rem 1.5rem;
  color: var(--dim);
  font-size: var(--sm);
  font-weight: 500;
}
</style>
