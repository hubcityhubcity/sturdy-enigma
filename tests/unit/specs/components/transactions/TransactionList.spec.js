import { shallowMount } from "@vue/test-utils"
import TransactionList from "src/components/transactions/TransactionList"

describe(`TransactionList`, () => {
  let wrapper

  const transactions = [
    {
      height: 1086769,
      fee: {
        amount: "3.7",
        denom: "atom"
      },
      group: "banking",
      key: "keyhash",
      memo: "(Sent via Lunie)",
      tiemstamp: new Date("2019-07-31"),
      liquidDate: null,
      type: "cosmos-sdk/MsgSend"
    },
    {
      height: 1086769,
      fee: {
        amount: "3.7",
        denom: "atom"
      },
      group: "banking",
      key: "keyhash2",
      memo: "(Sent via Lunie)",
      tiemstamp: new Date("1970-01-01"),
      liquidDate: null,
      type: "cosmos-sdk/MsgSend"
    },
    {
      height: 1086769,
      fee: {
        amount: "3.7",
        denom: "atom"
      },
      group: "banking",
      key: "keyhash3",
      memo: "(Sent via Lunie)",
      tiemstamp: new Date("2019-07-31"),
      liquidDate: null,
      type: "cosmos-sdk/MsgSend"
    }
  ]

  beforeEach(() => {
    wrapper = shallowMount(TransactionList, {
      propsData: {
        transactions: transactions,
        validators: {},
        address: "cosmos1"
      }
    })
  })

  it(`renders a list of TransactionItems`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
