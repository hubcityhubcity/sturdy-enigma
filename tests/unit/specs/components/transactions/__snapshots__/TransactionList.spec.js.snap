// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList renders a list of TransactionItems 1`] = `
<div>
  <div>
    <h3>
      January 19th
    </h3>
     
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
    <transactionitem-stub
      address="cosmos1"
      showmetadata="true"
      transaction="[object Object]"
      validators="[object Object]"
    />
  </div>
</div>
`;
